Spike: Make request to update order on merchant store asynchronous

reykjalin opened the pull request:
Spike. Main PR in 1223-gh-Automattic/woopay

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Automated tests should pass.
* Refer to platform checkout repo PR for detailed testing instructions.
* Make sure critical payment flows work as intended.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

Commit:
add woopay specific webhook handling for successful payments

Commit:
add WooPay info to order received page when coming from WooPay

Commit:
Merge branch 'develop' into add/no-update-request-from-woopay-to-merchant-store

reykjalin commented on includes/class-wc-payments-order-success-page.php:
```diff
@@ -14,11 +14,33 @@ class WC_Payments_Order_Success_Page {
+			return $order_id;
+		}
+
+		if ( ! empty( $_SERVER['HTTP_REFERER'] ) && wp_unslash( $_SERVER['HTTP_REFERER'] ) ?? '' !== 'http://woopay.test' ) {
```
I'm very unhappy with how we're detecting a request coming from WooPay, but can't really think of a different way to do this. We at least need some way to allow both the production URL as well as our local development environment URLs.

Then there's the question of how sites behind a reverse proxy or load balancer may potentially alter the value of this variable, so it doesn't feel very reliable. I'll be looking at ways to improve this, but would love to hear some feedback or ideas on this or other approaches.

reykjalin replied:
We could add a query parameter to the URL for the Order Received page redirect but then anyone can do that and trigger this code path. Not sure if that's the right approach here either.

reykjalin replied:
Resolving this since the approach has changed significantly since I initially worked on this.

reykjalin commented on includes/class-wc-payments-order-success-page.php:
```diff
@@ -14,11 +14,33 @@ class WC_Payments_Order_Success_Page {
+		}
+
+		if ( ! empty( $_SERVER['HTTP_REFERER'] ) && wp_unslash( $_SERVER['HTTP_REFERER'] ) ?? '' !== 'http://woopay.test' ) {
+			$order->set_payment_method( WC_Payment_Gateway_WCPay::GATEWAY_ID );
```
The payment method is empty (or some form of default) at this point so we must set it to WCPay to trigger the `woocommerce_thankyou_woocommerce_payments` hook later.

reykjalin commented on includes/class-wc-payments-order-success-page.php:
```diff
@@ -14,11 +14,33 @@ class WC_Payments_Order_Success_Page {
+		if ( ! empty( $_SERVER['HTTP_REFERER'] ) && wp_unslash( $_SERVER['HTTP_REFERER'] ) ?? '' !== 'http://woopay.test' ) {
+			$order->set_payment_method( WC_Payment_Gateway_WCPay::GATEWAY_ID );
+			$order->add_meta_data( 'is_woopay', true, true );
+			$order->set_status( 'pending' );
```
We set the status to `pending` because WooPay would have already processed a payment and as such the order should no longer be considered a draft.

We need to add checks here to make sure we don't mess with the status unless the order currently has the `draft` status. This is necessary to prevent potential race conditions if the webhooks are delivered before the customer reaches the order received page, however unlikely that is.

reykjalin commented on includes/class-wc-payments-order-success-page.php:
```diff
@@ -14,11 +14,33 @@ class WC_Payments_Order_Success_Page {
+
+		if ( ! empty( $_SERVER['HTTP_REFERER'] ) && wp_unslash( $_SERVER['HTTP_REFERER'] ) ?? '' !== 'http://woopay.test' ) {
+			$order->set_payment_method( WC_Payment_Gateway_WCPay::GATEWAY_ID );
+			$order->add_meta_data( 'is_woopay', true, true );
```
Adding this metadata is necessary to show the right information on the order received page. See the other hooks in this file for details on that; essentially a custom notice and the WooPay logo.

reykjalin commented on includes/class-wc-payments-order-success-page.php:
```diff
@@ -14,11 +14,33 @@ class WC_Payments_Order_Success_Page {
 		add_action( 'wp_enqueue_scripts', [ $this, 'enqueue_scripts' ] );
 	}
 
+	public function update_order_meta( $order_id ) {
```
The filter expects an ID to be returned, specifically the ID for the Order that will be displayed on the Order Received page.

reykjalin commented on includes/class-wc-payments-webhook-processing-service.php:
```diff
@@ -411,6 +413,67 @@ private function process_webhook_payment_intent_succeeded( $event_body ) {
+			$currency_store   = strtolower( get_option( 'woocommerce_currency' ) );
+			$currency_order   = strtolower( $order->get_currency() );
+			$currency_account = strtolower( $this->account->get_account_default_currency() );
+
+			// If the default currency for the store is different from the currency for the merchant's Stripe account,
+			// the conversion rate provided by Stripe won't make sense, so we should not attach it to the order meta data
+			// and instead we'll rely on the _wcpay_multi_currency_order_exchange_rate meta key for analytics.
+			if ( $currency_store !== $currency_account ) {
+				return;
+			}
+
+			if ( $currency_order !== $currency_account ) {
+				// We check that the currency used in the order is different than the one set in the WC Payments account
+				// to avoid requesting the charge if not needed.
+				$charge        = $this->payments_api_client->get_charge( $charge_id );
+				$exchange_rate = $charge['balance_transaction']['exchange_rate'] ?? null;
+				if ( isset( $exchange_rate ) ) {
+					$exchange_rate = WC_Payments_Utils::interpret_string_exchange_rate( $exchange_rate, $currency_order, $currency_account );
+					$order->update_meta_data( '_wcpay_multi_currency_stripe_exchange_rate', $exchange_rate );
+					$order->save_meta_data();
+				}
+			}
+
+			// update_order_status_from_intent.
+			switch ( $intent_status ) {
+				case 'succeeded':
+					$this->order_service->mark_payment_completed( $order, $intent_id, $intent_status, $charge_id );
+					break;
+				case 'processing':
+				case 'requires_capture':
+					$this->order_service->mark_payment_authorized( $order, $intent_id, $intent_status, $charge_id );
+					break;
+				case 'requires_action':
+				case 'requires_payment_method':
+					$this->order_service->mark_payment_started( $order, $intent_id, $intent_status, $charge_id );
+					break;
+				default:
+					Logger::error( 'Uncaught payment intent status of ' . $intent_status . ' passed for order id: ' . $order->get_id() );
+					break;
+			}
```
These are all direct copies of the `attach_intent_info_to_order`, `attach_exchange_info_to_order`, and `update_order_status_from_intent` functions from the `process_payment_for_order` function in the gateway. If we decide to go with this approach those functions should be moved to a shared class somewhere.

reykjalin replied:
I've since refactored these and added each copied function to the order service.

Commit:
Merge branch 'develop' into add/no-update-request-from-woopay-to-merchant-store

Commit:
Merge branch 'develop' into add/no-update-request-from-woopay-to-merchant-store

Commit:
Merge branch 'develop' into add/no-update-request-from-woopay-to-merchant-store

Commit:
Add class to handle WooPay order success page

Added Platform_Checkout_Order_Success_Page and moved previous iterations
of this branch there.

Commit:
Move order update function to service class

Commit:
Merge branch 'develop' into add/no-update-request-from-woopay-to-merchant-store

fashuvo commented:
**Size Change:** 0 B 

**Total Size:** 1.63 MB



<details><summary>ℹ️ <strong>View Unchanged</strong></summary>

| Filename | Size |
| :--- | :---: |
| `release/woocommerce-payments/assets/css/admin.css` | 913 B |
| `release/woocommerce-payments/assets/css/success.css` | 401 B |
| `release/woocommerce-payments/dist/blocks-checkout.css` | 1.77 kB |
| `release/woocommerce-payments/dist/blocks-checkout.js` | 27 kB |
| `release/woocommerce-payments/dist/checkout.css` | 1.85 kB |
| `release/woocommerce-payments/dist/checkout.js` | 25.9 kB |
| `release/woocommerce-payments/dist/index.css` | 35.8 kB |
| `release/woocommerce-payments/dist/index.js` | 627 kB |
| `release/woocommerce-payments/dist/multi-currency-analytics.js` | 8.22 kB |
| `release/woocommerce-payments/dist/multi-currency-switcher-block.js` | 71 kB |
| `release/woocommerce-payments/dist/multi-currency.css` | 14.7 kB |
| `release/woocommerce-payments/dist/multi-currency.js` | 109 kB |
| `release/woocommerce-payments/dist/onboarding.css` | 1.12 kB |
| `release/woocommerce-payments/dist/onboarding.js` | 38 kB |
| `release/woocommerce-payments/dist/order.js` | 299 B |
| `release/woocommerce-payments/dist/payment-gateways.css` | 1.2 kB |
| `release/woocommerce-payments/dist/payment-gateways.js` | 82.8 kB |
| `release/woocommerce-payments/dist/payment-request.js` | 22.4 kB |
| `release/woocommerce-payments/dist/platform-checkout.css` | 3.73 kB |
| `release/woocommerce-payments/dist/platform-checkout.js` | 76.6 kB |
| `release/woocommerce-payments/dist/settings.css` | 47.2 kB |
| `release/woocommerce-payments/dist/settings.js` | 148 kB |
| `release/woocommerce-payments/dist/subscription-edit-page.js` | 3.66 kB |
| `release/woocommerce-payments/dist/subscription-product-onboarding-modal.css` | 7.18 kB |
| `release/woocommerce-payments/dist/subscription-product-onboarding-modal.js` | 16.5 kB |
| `release/woocommerce-payments/dist/subscription-product-onboarding-toast.js` | 428 B |
| `release/woocommerce-payments/dist/subscriptions-empty-state.css` | 298 B |
| `release/woocommerce-payments/dist/subscriptions-empty-state.js` | 48.8 kB |
| `release/woocommerce-payments/dist/tos.css` | 236 B |
| `release/woocommerce-payments/dist/tos.js` | 22.7 kB |
| `release/woocommerce-payments/dist/upe_checkout.css` | 1.85 kB |
| `release/woocommerce-payments/dist/upe_checkout.js` | 37.2 kB |
| `release/woocommerce-payments/dist/upe-blocks-checkout.css` | 1.77 kB |
| `release/woocommerce-payments/dist/upe-blocks-checkout.js` | 43.4 kB |
| `release/woocommerce-payments/includes/subscriptions/assets/css/plugin-page.css` | 633 B |
| `release/woocommerce-payments/includes/subscriptions/assets/js/plugin-page.js` | 355 B |
| `release/woocommerce-payments/vendor/automattic/jetpack-admin-ui/src/css/jetpack-icon.css` | 224 B |
| `release/woocommerce-payments/vendor/automattic/jetpack-assets/build/i18n-loader.js` | 2.43 kB |
| `release/woocommerce-payments/vendor/automattic/jetpack-assets/src/js/i18n-loader.js` | 1.01 kB |
| `release/woocommerce-payments/vendor/automattic/jetpack-identity-crisis/babel.config.js` | 160 B |
| `release/woocommerce-payments/vendor/automattic/jetpack-identity-crisis/build/index.css` | 2.27 kB |
| `release/woocommerce-payments/vendor/automattic/jetpack-identity-crisis/build/index.js` | 13.3 kB |
| `release/woocommerce-payments/vendor/automattic/jetpack-identity-crisis/build/index.rtl.css` | 2.27 kB |
| `release/woocommerce-payments/vendor/automattic/jetpack-tracking/src/js/tracks-ajax.js` | 789 B |
| `release/woocommerce-payments/vendor/automattic/jetpack-tracking/src/js/tracks-callables.js` | 925 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/css/about.css` | 1.2 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/css/admin-order-statuses.css` | 403 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/css/admin.css` | 3.43 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/css/checkout.css` | 299 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/css/modal.css` | 742 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/css/view-subscription.css` | 572 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/css/wcs-upgrade.css` | 411 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/admin-pointers.js` | 544 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/admin.js` | 8.97 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/jstz.js` | 6.8 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/jstz.min.js` | 3.83 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/meta-boxes-coupon.js` | 544 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/meta-boxes-subscription.js` | 2.35 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/moment.js` | 22.1 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/moment.min.js` | 11.6 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/payment-method-restrictions.js` | 1.29 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/wcs-meta-boxes-order.js` | 502 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/frontend/payment-methods.js` | 355 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/frontend/single-product.js` | 429 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/frontend/view-subscription.js` | 1.38 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/frontend/wcs-cart.js` | 387 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/modal.js` | 1.1 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/wcs-upgrade.js` | 1.27 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/build/index.css` | 392 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/build/index.js` | 3.06 kB |

</details>



<a href="https://github.com/preactjs/compressed-size-action"><sub>compressed-size-action</sub></a>

Commit:
Merge branch 'develop' into add/no-update-request-from-woopay-to-merchant-store

Commit:
Merge branch 'develop' into add/no-update-request-from-woopay-to-merchant-store

Commit:
Merge branch 'develop' into add/no-update-request-from-woopay-to-merchant-store

Commit:
use new order service methods everywhere appropriate

Commit:
update unit tests

reykjalin commented on tests/unit/test-class-wc-payment-gateway-wcpay-process-payment.php:
```diff
@@ -340,6 +335,19 @@ public function test_intent_status_requires_capture() {
 		// Arrange: Create an order to test with.
 		$mock_order = $this->createMock( 'WC_Order' );
 
+		// Arrange: "data store" for when order is cloned in order service.
+		$mock_data_store = new class( $mock_order ) {
+			// phpcs:ignore
+			public function __construct( $mock_order ) { $this->mock_order = $mock_order; }
+			// phpcs:ignore
+			public function read( $unused ) { return $this->mock_order; }
+		};
+
+		// Arrange: Set a good return value for `get_data_store()`.
+		$mock_order
+			->method( 'get_data_store' )
+			->willReturn( $mock_data_store );
```
The approach to this is a little bit weird, but I couldn't find a better way to do this.

We need to mock this on the order because the tests were changed to use an actual instance of the order service instead of a mocked instance.

cesarcosta99 replied:
I commented about this in my previous comment. I think this could be avoided by actually mocking the order service?

Commit:
Process orders with downloads on success page

Commit:
Merge branch 'develop' into add/no-update-request-from-woopay-to-merchant-store

Commit:
Handle new event type for WooPay intent success

reykjalin commented on includes/platform-checkout/class-platform-checkout-order-success-page.php:
```diff
@@ -0,0 +1,193 @@
+			return;
+		}
+
+		$payment_method = isset( $_GET['wc_payment_method'] ) ? wc_clean( wp_unslash( $_GET['wc_payment_method'] ) ) : '';
+		if ( self::GATEWAY_ID !== $payment_method ) {
+			return;
+		}
+
+		$from_woopay = isset( $_GET['from_woopay'] ) ? filter_var( wc_clean( wp_unslash( $_GET['from_woopay'] ) ), FILTER_VALIDATE_BOOL ) : false;
+		if ( ! $from_woopay ) {
+			return;
+		}
+
+		$order_id  = isset( $_GET['order_id'] ) ? wc_clean( wp_unslash( $_GET['order_id'] ) ) : '';
+		$last_4    = isset( $_GET['last4'] ) ? wc_clean( wp_unslash( $_GET['last4'] ) ) : '';
+		$intent_id = isset( $_GET['payment_intent'] ) ? wc_clean( wp_unslash( $_GET['payment_intent'] ) ) : '';
```
These accesses to the `$_GET` variable all throw a linting error because we're not validating a nonce. Since the redirect should be coming from WooPay I'm not even sure if there is a nonce to validate so I'm not really sure how to handle that at the moment.

Maybe we need to send a nonce with the WooPay session init request and include it on the redirect for verification?

reykjalin replied:
It looks like passing a nonce through the WooPay process is the way to go (see https://github.com/Automattic/woocommerce-payments/pull/4604/commits/70466f1609e95c900a4624c35af374e62a0394da).

Commit:
add nonce verification for order success redirect

Commit:
add changelog entry

Commit:
fix issues reported by psalm

Commit:
Merge branch 'develop' into add/no-update-request-from-woopay-to-merchant-store

Commit:
Merge branch 'develop' into add/no-update-request-from-woopay-to-merchant-store

Commit:
Add support for subscriptions bypass

We're not yet adding support for subscriptions with the new optimized
approach so we need to add a proper bypass when applicable.

Commit:
Add support for digital downloads bypass

We're not yet adding support for digital downloads with the new
optimized approach so we need to add a proper bypass when applicable.

Commit:
Merge branch 'develop' into add/no-update-request-from-woopay-to-merchant-store

Commit:
Run post checkout action to trigger Tracks events

Not sure if this is actually the best way to do this. Might be better to
instantiate a Platform_Tracks client and run the function there
manually. Definitely nervous about side effects from running the action
when the customer isn't on the checkout page.

Commit:
Trigger tracks events without side effects

Using the action like before would have unintended side effects that we
want to avoid so we trigger the Tracks events manually instead.

Commit:
fix psalm errors and warnings

Commit:
track checkouts coming in from WooPay

Commit:
consolidate woopay and wcpay intent webhook handling

Commit:
Merge branch 'develop' into add/no-update-request-from-woopay-to-merchant-store

reykjalin commented on includes/class-platform-checkout-tracker.php:
```diff
@@ -282,12 +282,16 @@ public function checkout_start() {
-			[
-				'source' => isset( $_SERVER['HTTP_X_WCPAY_PLATFORM_CHECKOUT_USER'] ) ? 'platform' : 'standard',
-			]
-		);
+		if ( isset( $_SERVER['HTTP_X_WCPAY_PLATFORM_CHECKOUT_USER'] ) ) {
+			$this->platform_checkout_order_processed();
+		} else {
+			$this->maybe_record_event(
+				'order_checkout_complete',
+				[
+					'source' => 'standard',
+				]
+			);
+		}
 	}
```
@brianyu28 I'd love it if you could take a look at these changes to the tracking function (and the new `platform_checkout_processed` function) just to check if this looks right 🙏

brianyu28 replied:
:+1: The tracking function changes look good to me.

cesarcosta99 commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -1164,9 +1164,9 @@ public function process_payment_for_order( $cart, $payment_information, $additio
-		$this->update_order_status_from_intent( $order, $intent_id, $status, $charge_id );
+		$this->order_service->attach_intent_info_to_order( $order, $intent_id, $status, $payment_method, $customer_id, $charge_id, $currency );
+		$this->order_service->attach_exchange_info_to_order( $order, $charge_id, $this->account );
+		$this->order_service->update_order_status_from_intent( $order, $intent_id, $status, $charge_id );
```
Good call moving all of this logic to the order service 👍

cesarcosta99 commented on includes/platform-checkout/class-platform-checkout-order-success-page.php:
```diff
@@ -0,0 +1,228 @@
+	}
+
+	/**
+	 * Doc comment.
```
🤏 Nit: we might want to add something relevant as a comment 😆

cesarcosta99 commented on includes/platform-checkout/class-platform-checkout-order-success-page.php:
```diff
@@ -0,0 +1,228 @@
+			return;
+		}
+
+		$payment_method = isset( $_GET['wc_payment_method'] ) ? wc_clean( wp_unslash( $_GET['wc_payment_method'] ) ) : '';
+		if ( self::GATEWAY_ID !== $payment_method ) {
+			return;
+		}
+
+		$from_woopay = isset( $_GET['from_woopay'] ) ? filter_var( wc_clean( wp_unslash( $_GET['from_woopay'] ) ), FILTER_VALIDATE_BOOL ) : false;
+		if ( ! $from_woopay ) {
+			return;
+		}
```
Wouldn't suffice check whether the payment method is `woopay`? I can't think on other ways a user would land on the success page with `woopay` as payment method besides coming from WooPay itself.

If my reasoning is correct, we could discard this `from_woopay` request parameter.

cesarcosta99 commented on tests/unit/subscriptions/test-class-wc-payments-invoice-service.php:
```diff
@@ -43,7 +50,8 @@ public function set_up() {
 		$this->mock_product_service = $this->createMock( WC_Payments_Product_Service::class );
 		$this->mock_gateway         = $this->createMock( WC_Payment_Gateway_WCPay::class );
-		$this->invoice_service      = new WC_Payments_Invoice_Service( $this->mock_api_client, $this->mock_product_service, $this->mock_gateway );
+		$this->order_service        = new WC_Payments_Order_Service( $this->mock_api_client );
```
We could mock the order service here unless there's a reason not to. That's how it's done for other services, like the product one.

cesarcosta99 reviewed (commented):
Leaving general feedback while we address possible changes in the approach.
